### ISSUE TITLE
What is the recommended solution to clear spin-related cache files in the /tmp directory?

### ISSUE THREAD

**Issue by qtfkwk** [fermyon/spin#1698]
Loving spin otherwise btw!

I've been using spin locally via `spin watch` during development of an app using 3 components (the spin-fileserver and 2 custom ones) on a Debian Linux system for weeks. Noticed today my `/tmp` partition was nearly out of space. Seeing thousands of directories called `.tmpRANDOM` (where `RANDOM` is a random 6-digit alphanumeric string) with `spin.lock` (so I know they're related to spin) in `/tmp` totaling to hundreds of GB!

Not observing any subcommand in the `spin` CLI addressing cleaning up cache, etc...

Presuming I can kill `spin watch` and any other instances of spin, and delete all of these via `fd '^spin.lock$' /tmp |sed 's/\/spin\.lock$//' |xargs rm -rf`?

Is there any other solution you would recommend?

Are there any methods to prevent this behavior... such as... spin automatically deleting cache files in /tmp when it exits?

Thanks!

---

Spin version (`spin --version`): `spin 1.4.1 (e0bd911 2023-07-12)`

Installed plugin versions (`spin plugins list --installed`)

```
cloud 0.1.2 [installed]
js2wasm 0.5.1 [installed]
py2wasm 0.3.0 [installed]
```

**Comment by itowlson** [comment]
Thanks for flagging this.  Spin is meant to delete its temporary directory on exit - I'm not yet sure what's going on here but a possible theory is that, when `spin watch` detects a change, it is terminating `spin up` in a way that bypasses the cleanup.  I'll look into this.

In the meantime, yes, it is safe to delete these temporary directories when Spin is not running.  Sorry for the inconvenience!

**Comment by itowlson** [comment]
I've confirmed that this is a `spin watch` issue.  Running `spin up` and Ctrl+C-ing it deletes the temp directory; running `spin watch` creates a new temp directory each time the application is restarted, and Ctrl+C-ing it doesn't delete any of these directories.

`spin watch` uses `watchexec`, which controls process termination and restart - that is, we do not directly control the code that actually terminates `spin up` - and as far as I can tell, it does this by sending SIGKILL, which is handled at the OS level and doesn't give `spin up` a chance to tidy up behind itself.  I can't see a way for us to override this strategy.  Unless I'm missing something, this gives us two ways forward:

1. Change `spin up` so that the temp directory is transient at the OS level.  This would make cleanup safer in general.  I believe on Unix this can be done by holding a file descriptor to the temp directory while unlinking it at the filesystem level, but I am not very au fait with this... and I am not sure if there is a strategy that would be portable to Windows.
2. Change `spin watch` to watch files manually, using `notify`/`notify_debouncer_mini`, and manage rebuilds/reloads ourselves.  (This could even use the build and up structures in-process instead of shelling out to them.)  This would be fully portable, and would give us more visibility and control going forward, but would involve significant rework.

@calebschoepp thoughts or better ideas?

**Comment by lann** [comment]
It looks like `watchexec` _does_ support sending other signals. A snippet from its repo that seems likely to do what we want: https://github.com/watchexec/watchexec/blob/96bf3d231e6312beda9a589b52e8b64138514293/crates/cli/src/config/runtime.rs#L174-L177
```

**Comment by calebschoepp** [comment]
@itowlson did you repro this on Debian linux like the OP? I've run a lot of `spin watch` on my Mac and at a brief glance I see nothing my `/tmp` directory so I want to determine if this is only an issue on Debian Linux.

> 1. Change spin up so that the temp directory is transient at the OS level. This would make cleanup safer in general. I believe on Unix this can be done by holding a file descriptor to the temp directory while unlinking it at the filesystem level, but I am not very au fait with this... and I am not sure if there is a strategy that would be portable to Windows.

Can you explain what you mean by transient at the OS level in simpler terms. I'm not quite following.

> 2. Change spin watch to watch files manually, using notify/notify_debouncer_mini, and manage rebuilds/reloads ourselves. (This could even use the build and up structures in-process instead of shelling out to them.) This would be fully portable, and would give us more visibility and control going forward, but would involve significant rework.

Definitely agree on some of the advantages here that you've listed. That being said I ended up not going in this direction (and I spent some significant time trying) because we would have had to handle so many cross OS details ourselves.

I think what Lann linked is the right direction to explore. In our implementation we are only using `Outcome::Stop` which after some digging appears to be backed by `SIGKILL` (hence the issue we're having). In Lann's example they use `Outcome::Signal()` to be able to send something other than `SIGKILL`. We could send something like `SIGINT` instead. The catch is that it looks like we'll have to use macros to keep using `Outcome::Stop` on Windows since the Unix signals aren't supported.

**Comment by itowlson** [comment]
@lann Yeah the CLI does it.  I did quite a bit of gazing at it trying to figure out how to set the equivalent of `--stop-signal` in the library.  But I think it's not a direct property, it's an Outcome we choose.  So we can maybe change https://github.com/fermyon/spin/blob/45b14fb48eada7c87d38b570945210e8a9d2a8f5/src/watch_state.rs#L82 or something like that to send `Outcome::Signal(...)` instead of `Outcome::Stop`.  @calebschoepp does that sound right?

@calebschoepp Yes, I am using Ubuntu on WSL.

By "transient at the OS level" I was thinking of the way Unix allows you to create a file, grab a descriptor / handle / ID / something, then delete the file, and the in-memory thing means the file continues to exist _as long as the in-memory thing does_.  My understanding (which I look forward to having corrected) is that Unix files are essentially reference counted, with each directory entry being a reference (and there can be several directory entries for the same file because links).  Only when all entries/links are deleted does the actual file get destroyed (hence Unix calling it "unlink" rather than "delete").  And process holding an open reference counts as a link.  So if a process creates a file on disk (1 link), opens it (2 links), then deletes the file on disk (1 link), the file still exists!  But when the process closes the handle it goes down to 0 links and gets removed.  Sorry I am explaining this very meanderingly but hopefully it rings bells!

But yeah if we change the Outcome::Stop to Outcome::Signal(SIGINT) then the above rambly paragraph becomes irrelevant!

**Comment by calebschoepp** [comment]
> or something like that to send Outcome::Signal(...) instead of Outcome::Stop. @calebschoepp does that sound right?
Yes that is what I'm advocating. With the caveat that I think we need to keep using `Outcome::Stop` for Windows.

> By "transient at the OS level"...
Thanks that helped.

**Comment by itowlson** [comment]
Changing `Stop` to `Signal(Interrupt)` fixed the thing where it didn't clean up temp directories, but introduces a rather bigger problem where it sometimes continues to serve the previous build after a change.  It looks like in these cases the build is not completing before the app is started (visually, `Finished building all Spin components` is not printed before the `up` output such as `Logging component stdio to...` starts appearing).  This was _not_ happening with the `Stop` outcome.  Continuing to investigate.

**Comment by itowlson** [comment]
In these cases `spin build` starts but does not exit (normally) and does not complete the build.  This is implied by some test logging, and is confirmed by running `spin up` - the .`wasm` file remains the old one.  It is as if `build` had been killed before it could write out the updated binary.

**Comment by itowlson** [comment]
Oh!  I think there is an overlap.  Here is a trace.  Here, Spin is modified to print `SPIN STARTING` as soon as it enters `main`, and `SPIN EXITING` as the last thing it does before returning from `main`.

```
*** handling effect SourceChange in current state Running
*** SPIN EXITING OKAY: Args { inner: ["/home/ivan/github/spin/target/debug/spin", "trigger", "http"] }
*** SPIN STARTING: Args { inner: ["/home/ivan/github/spin/target/debug/spin", "build", "-f", "spin.toml"] }
*** SPIN EXITING OKAY: Args { inner: ["/home/ivan/github/spin/target/debug/spin", "up", "-f", "spin.toml"] }
Building component dioxus-test with `cargo build --target wasm32-wasi --release`
    Blocking waiting for file lock on package cache
    Blocking waiting for file lock on package cache
*** handling effect ChildProcessCompleted in current state Building
*** SPIN STARTING: Args { inner: ["/home/ivan/github/spin/target/debug/spin", "up", "-f", "spin.toml"] }
*** SPIN STARTING: Args { inner: ["/home/ivan/github/spin/target/debug/spin", "trigger", "http"] }
```

The `build` process prints `STARTING` before the `up` process prints `EXITING`.  We rightly get `ChildProcessCompleted` for the `up` process.  But `watch` has already moved into the `Building` state and kicked off `spin build`.  And so `watch` mistakenly thinks the completion is for the `build` process... which causes it to go into the `Running` state and (I _think_) stop the current process (i.e. kill `spin build) and do a `Start` outcome (which I think in `Running` kicks off `spin up`).

So `watch` needs to change to _actually wait for `spin up` to fully exit_, before going into its build routine.  It currently seems to switch and kick off the build immediately on the source file change, which worked when process killing was instantaneous, but is no longer safe if process exit waits for cleanup code for run.
